### PR TITLE
Move Cypher expressions to own module

### PR DESCRIPTION
**Pull request by fickludd**
The cypher compiler is a very complex piece of machinery, which is hard to understand because it's divided into only a few modules, which depend heavily on one another. 

Here cypher expressions are isolated into their own module. The same expressions classes are used directly in the AST, and then kept around throughout compilation and into the logical plan. In 3.3 they lived in the `frontend` module, and were heavily entangled in semantic analysis. However, most users of expressions are not concerned with semantic analysis, and should therefore not have to care about changes to this specific part of the compilation process. With the PR, the expressions have their own module, which in turn only depends on the new leaf module `util`.

To achieve this split, I had to

1. Move cypher global helper classes into a new leaf `util` module
2. Untangle expressions, patterns and functions from semantic analysis
3. Perform the actual move, together with all associated import updates

For reviewing, I would recommend to checkout the branch, and focus on the two new modules, as well as the changes to semantic analysis. Before review, this PR needs to be rebased on `3.4` once more when all the `3.3` cypher modules have been updated to their `3.4` variants.

### CODE REVIEW

**Comment by fickludd** [comment]
@henriknyman @thobe. Rebased and ready for you. If acceptable I would prefer to merge this soon, as rebasing all the imports is very annoying and time consuming. It would be nice to make necessary improvements in a follow up PR.

**Comment by thobe** [comment]
Error in module `org.neo4j:neo4j-enterprise-kernel`: Could not find artifact `org.neo4j:neo4j-cypher-util-3.4` - _looks like you removed a module but not all references to it_.

File: community/cypher/cypher-compiler-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/CardinalityCostModel.scala [hunk 1]
```diff
@@ -20,16 +20,16 @@
...
+import org.neo4j.cypher.internal.v3_4.expressions.{HasLabels, Property}
 
 object CardinalityCostModel extends CostModel {
   def VERBOSE = java.lang.Boolean.getBoolean("CardinalityCostModel.VERBOSE")
 
   private val DEFAULT_COST_PER_ROW: CostPerRow = 0.1
-  private val PROBE_BUILD_COST: CostPerRow = 3.1
-  private val PROBE_SEARCH_COST: CostPerRow = 2.4
+  private val PROBE_BUILD_COST: CostPerRow = 19.4
+  private val PROBE_SEARCH_COST: CostPerRow = 11.5
```
Comment (henriknyman) [inline community/cypher/cypher-compiler-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/CardinalityCostModel.scala]: Is the cost model supposed to be updated in this PR?

File: community/cypher/cypher-compiler-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/NamespacerTest.scala [hunk 2]
```diff
@@ -82,28 +85,6 @@ class NamespacerTest extends CypherFunSuite with AstConstructionTestSupport {
...
-    val original =
-      """FROM GRAPH AT '/test/graph2' AS myGraph
-        |WITH 1 AS a GRAPHS myGraph
-        |MATCH (n:Person)
-        |WITH n GRAPHS GRAPH AT 'foo' AS fooG >> myGraph AS barG
-        |RETURN n.name AS name GRAPHS fooG, barG
-      """.stripMargin
-
-    assertRewritten(original, original, SemanticFeature.MultipleGraphs)
-  }
```
Comment (henriknyman) [inline community/cypher/cypher-compiler-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/ast/rewriters/NamespacerTest.scala]: Are these tests not needed anymore?

**Review by henriknyman — COMMENTED** [review]
Review in progress...

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/helpers.scala [hunk 3]
```diff
@@ -17,15 +17,15 @@
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.cypher.internal.compatibility.v3_4
+package org.neo4j.cypher.internal.compatibility.v3_2
```
Comment (henriknyman) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/helpers.scala]: Is v3_2 intentional here?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_1/helpers.scala [hunk 4]
```diff
@@ -52,12 +52,12 @@ object helpers {
...
-          case v3_1Phase.LOGICAL_PLANNING => v3_4Phase.LOGICAL_PLANNING
-          case v3_1Phase.PARSING => v3_4Phase.PARSING
-          case v3_1Phase.PIPE_BUILDING => v3_4Phase.PIPE_BUILDING
-          case v3_1Phase.SEMANTIC_CHECK => v3_4Phase.SEMANTIC_CHECK
+          case v3_1Phase.AST_REWRITE => v3_3Phase.AST_REWRITE
+          case v3_1Phase.CODE_GENERATION => v3_3Phase.CODE_GENERATION
+          case v3_1Phase.LOGICAL_PLANNING => v3_3Phase.LOGICAL_PLANNING
+          case v3_1Phase.PARSING => v3_3Phase.PARSING
+          case v3_1Phase.PIPE_BUILDING => v3_3Phase.PIPE_BUILDING
+          case v3_1Phase.SEMANTIC_CHECK => v3_3Phase.SEMANTIC_CHECK
```
Comment (henriknyman) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_1/helpers.scala]: Is this change on purpose or  should it be v3_4?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v2_3/ExecutionResultWrapper.scala [hunk 5]
```diff
@@ -155,9 +157,9 @@ class ExecutionResultWrapper(val inner: InternalExecutionResult, val planner: Pl
     case Arguments.EstimatedRows(value) => InternalPlanDescription3_3.Arguments.EstimatedRows(value)
     case Arguments.ExpandExpression(from, relName, relTypes, to, direction, varLength) =>
       val dir3_3 = direction match {
-        case INCOMING => v3_4.SemanticDirection.INCOMING
-        case OUTGOING => v3_4.SemanticDirection.OUTGOING
-        case BOTH => v3_4.SemanticDirection.BOTH
+        case INCOMING => SemanticDirection.INCOMING
+        case OUTGOING => expressions.SemanticDirection.OUTGOING
```
Comment (henriknyman) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v2_3/ExecutionResultWrapper.scala]: They could all be `expression.`... (not important)

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/planDescription/LogicalPlan2PlanDescription.scala [hunk 6]
```diff
@@ -19,22 +19,24 @@
...
+import org.neo4j.cypher.internal.v3_4.expressions.{Expression => ASTExpression}
 
 object LogicalPlan2PlanDescription extends ((LogicalPlan, PlannerName) => InternalPlanDescription) {
 
   override def apply(input: LogicalPlan,
                      plannerName: PlannerName): InternalPlanDescription = {
     val readOnly = input.solved.readOnly
     new LogicalPlan2PlanDescription(readOnly).create(input)
-      .addArgument(Version("CYPHER 3.4"))
+      .addArgument(Version("CYPHER 3.3"))
```
Comment (henriknyman) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/planDescription/LogicalPlan2PlanDescription.scala]: Why 3.3 ?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/ExecutionEngineIT.scala [hunk 7]
```diff
@@ -77,11 +77,11 @@ class ExecutionEngineIT extends CypherFunSuite with GraphIcing {
     plan2.getArguments.get("planner-impl") should equal(CostBasedPlannerName.default.name)
   }
 
-  test("by default when using cypher 3.4 some queries should default to COST") {
+  test("by default when using cypher 3.3 some queries should default to COST") {
```
Comment (henriknyman) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/ExecutionEngineIT.scala]: Why 3.3 in this file?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/QueryCachingTest.scala [hunk 8]
```diff
@@ -73,9 +73,9 @@ class QueryCachingTest extends CypherFunSuite with GraphDatabaseTestSupport with
         val actual = cacheListener.trace
         val expected = List(
           s"cacheFlushDetected",
-          s"cacheMiss: CYPHER 3.4 $query",
-          s"cacheHit: CYPHER 3.4 $query",
-          s"cacheHit: CYPHER 3.4 $query")
+          s"cacheMiss: CYPHER 3.3 $query",
+          s"cacheHit: CYPHER 3.3 $query",
+          s"cacheHit: CYPHER 3.3 $query")
```
Comment (henriknyman) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/QueryCachingTest.scala]: Why 3.3?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/planDescription/LogicalPlan2PlanDescriptionTest.scala [hunk 9]
```diff
@@ -47,33 +49,33 @@ class LogicalPlan2PlanDescriptionTest extends CypherFunSuite with TableDrivenPro
...
 
       , Expand(lhsLP, IdName("a"), SemanticDirection.OUTGOING, Seq.empty, IdName("a"), IdName("r1"), ExpandInto)(113) ->
         PlanDescriptionImpl(id, "Expand(Into)", SingleChild(lhsPD), Seq(ExpandExpression("a", "r1", Seq.empty, "a", SemanticDirection.OUTGOING, 1, Some(1)),
-                                                                        EstimatedRows(113), Version("CYPHER 3.4"), Planner("COST"), PlannerImpl("IDP")), Set("a", "r1"))
+                                                                        EstimatedRows(113), Version("CYPHER 3.3"), Planner("COST"), PlannerImpl("IDP")), Set("a", "r1"))
 
       , NodeHashJoin(Set(IdName("a")), lhsLP, rhsLP)(2345) ->
-        PlanDescriptionImpl(id, "NodeHashJoin", TwoChildren(lhsPD, rhsPD), Seq(KeyNames(Seq("a")), EstimatedRows(2345), Version("CYPHER 3.4"), Planner("COST"), PlannerImpl("IDP")), Set("a", "b"))
+        PlanDescriptionImpl(id, "NodeHashJoin", TwoChildren(lhsPD, rhsPD), Seq(KeyNames(Seq("a")), EstimatedRows(2345), Version("CYPHER 3.3"), Planner("COST"), PlannerImpl("IDP")), Set("a", "b"))
```
Comment (henriknyman) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/planDescription/LogicalPlan2PlanDescriptionTest.scala]: Why 3.3?

File: community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/Namespacer.scala [hunk 10]
```diff
@@ -56,33 +59,17 @@ object Namespacer extends Phase[BaseContext, BaseState, BaseState] {
   private def returnAliases(statement: Statement): Set[Ref[Variable]] =
     statement.treeFold(Set.empty[Ref[Variable]]) {
 
-      case With(_, _, GraphReturnItems(_, items), _, _, _, _) =>
-        val gVars = extractGraphVars(items)
-        acc => (acc ++ gVars, Some(identity))
-
```
Comment (henriknyman) [inline community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/rewriters/Namespacer.scala]: Is this intentionally removed?

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/ExpressionTest.scala [hunk 11]
```diff
@@ -16,59 +16,16 @@
...
-    result.errors should have size 1
-    result.errors.head.position should equal(expression.position)
-    expression.types(result.state) shouldBe empty
-
-    assert(result.errors.size === 1)
-    assert(result.errors.head.position === expression.position)
-    assert(result.errors.head.msg == "Type mismatch: lhs was String yet rhs was Integer or Node")
-    assert(expression.types(result.state).isEmpty)
-  }
-
```
Comment (henriknyman) [inline community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/ExpressionTest.scala:71]: Are these tests not needed anymore?

File: community/cypher/pom.xml [hunk 12]
```diff
@@ -103,7 +137,7 @@
         <plugin>
           <groupId>org.scalastyle</groupId>
           <artifactId>scalastyle-maven-plugin</artifactId>
-          <version>1.0.0</version>
+          <version>0.4.0</version>
```
Comment (henriknyman) [inline community/cypher/pom.xml]: 1.0.0 was no good?

File: community/cypher/pom.xml [hunk 13]
```diff
@@ -49,13 +54,43 @@
...
+        <configuration>
+          <args>
+              <arg>-Xmax-classfile-name</arg>
+              <arg>100</arg>
+              <!-- arg>-deprecation</arg -->
+          </args>
+          <jvmArgs>
+            <jvmArg>-Xms64m</jvmArg>
+            <jvmArg>-Xmx1024m</jvmArg>
+          </jvmArgs>
```
Comment (henriknyman) [inline community/cypher/pom.xml]: Does this make compilation faster?

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/QueryPlanCompactionAcceptanceTest.scala [hunk 14]
```diff
@@ -829,7 +829,7 @@ class QueryPlanCompactionAcceptanceTest extends ExecutionEngineFunSuite with Que
     val result = executeWithCostPlannerAndCompiledRuntimeOnly(query)
     result should havePlanLike(
       """
-        |Compiler CYPHER 3.4
+        |Compiler CYPHER 3.3
```
Comment (henriknyman) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/QueryPlanCompactionAcceptanceTest.scala]: Why 3.3?

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala [hunk 15]
```diff
@@ -359,31 +359,31 @@ class SemanticErrorAcceptanceTest extends ExecutionEngineFunSuite {
     val mess = "Can't use aggregating expressions inside of expressions executing over lists"
     executeAndEnsureError(
       "MATCH (n) RETURN ALL(x in [1,2,3,4,5] WHERE count(*) = 0)",
-      s"$mess (line 1, column 27 (offset: 26))")
+      s"$mess (line 1, column 45 (offset: 44))")
```
Comment (henriknyman) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala:362]: The column number seem to be more correct now?

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/codegen/CodeGenConfiguration.scala [hunk 16]
```diff
@@ -43,13 +43,13 @@ case object ByteCodeMode extends CodeGenMode
  *
   * @param mode The mode of code generation
   * @param showSource if `true` source code is stored and returned
-  * @param packageName The name of the v3_4 the produced code should belong to
+  * @param packageName The name of the v3_3 the produced code should belong to
```
Comment (henriknyman) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/codegen/CodeGenConfiguration.scala]: Should be v3_4?

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/codegen/CodeGenConfiguration.scala [hunk 17]
```diff
@@ -43,13 +43,13 @@ case object ByteCodeMode extends CodeGenMode
...
   */
 case class CodeGenConfiguration(mode: CodeGenMode = CodeGenMode.default,
                                 showSource: Boolean = false,
                                 showByteCode: Boolean = false,
                                 saveSource: Option[Path] = None,
-                                packageName: String = "org.neo4j.cypher.internal.compiler.v3_4.generated"
+                                packageName: String = "org.neo4j.cypher.internal.compiler.v3_3.generated"
```
Comment (henriknyman) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/runtime/compiled/codegen/CodeGenConfiguration.scala]: v3_4?

File: enterprise/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiled_runtime/spi/v3_4/GeneratedMethodStructureTest.scala [hunk 18]
```diff
@@ -17,26 +17,28 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program. If not, see <http://www.gnu.org/licenses/>.
  */
-package org.neo4j.cypher.internal.compiled_runtime.spi.v3_4
+package org.neo4j.cypher.internal.compiled_runtime.spi.v3_3
```
Comment (henriknyman) [inline enterprise/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/compiled_runtime/spi/v3_4/GeneratedMethodStructureTest.scala]: This should be v3_4?

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v2_3/ExecutionResultWrapper.scala [hunk 19]
```diff
@@ -36,6 +36,8 @@ import org.neo4j.cypher.internal.frontend.v2_3.SemanticDirection.{BOTH, INCOMING
 import org.neo4j.cypher.internal.frontend.v2_3.notification.{InternalNotification, LegacyPlannerNotification, PlannerUnsupportedNotification, RuntimeUnsupportedNotification, _}
 import org.neo4j.cypher.internal.frontend.v2_3.{InputPosition => InternalInputPosition}
 import org.neo4j.cypher.internal.frontend.{v2_3, v3_4}
+import org.neo4j.cypher.internal.v3_4.expressions
+import org.neo4j.cypher.internal.v3_4.expressions.SemanticDirection
```
Comment (henriknyman) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v2_3/ExecutionResultWrapper.scala]: This file has some names called xxx3_3 that could be renamed to xxx3_4

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/helpers.scala [hunk 20]
```diff
@@ -17,15 +17,15 @@
...
 
-import org.neo4j.cypher.internal.compiler.v3_3.{CypherCompilerConfiguration => CypherCompilerConfiguration3_3}
+import org.neo4j.cypher.internal.aux.v3_4.{InputPosition, InternalException}
+import org.neo4j.cypher.internal.compiler.v3_4.{CypherCompilerConfiguration => CypherCompilerConfiguration3_3}
```
Comment (henriknyman) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_4/helpers.scala]: This file has some names called xxx3_3 that could be renamed to xxx3_4

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_1/helpers.scala [hunk 21]
```diff
@@ -20,14 +20,14 @@
 package org.neo4j.cypher.internal.compatibility.v3_1
 
 import org.neo4j.cypher.InternalException
+import org.neo4j.cypher.internal.aux.v3_4.InputPosition
 import org.neo4j.cypher.internal.compiler.v3_1
 import org.neo4j.cypher.internal.compiler.v3_1.CompilationPhaseTracer.{CompilationPhaseEvent, CompilationPhase => v3_1Phase}
 import org.neo4j.cypher.internal.compiler.v3_1.{CypherCompilerConfiguration => CypherCompilerConfiguration3_1}
+import org.neo4j.cypher.internal.frontend.v3_4.phases.CompilationPhaseTracer.{CompilationPhase => v3_3Phase}
```
Comment (henriknyman) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_1/helpers.scala]: This file has some names called xxx3_3 that could be renamed to xxx3_4

File: community/cypher/expressions-3.4/pom.xml [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+        <groupId>net.alchim31.maven</groupId>
+        <artifactId>scala-maven-plugin</artifactId>
+        <configuration>
+          <scalaVersion>${scala.version}</scalaVersion>
+          <scalaCompatVersion>${scala.binary.version}</scalaCompatVersion>
+          <args>
+            <arg>-Xmax-classfile-name</arg>
+            <arg>100</arg>
+            <arg>-Xlint</arg>
+          </args>
```
Comment (henriknyman) [inline community/cypher/expressions-3.4/pom.xml]: Maybe check that this is in sync with updates of scala-maven-plugin in other pom-files?

File: community/cypher/frontend-3.4/pom.xml [hunk 23]
```diff
@@ -38,6 +38,21 @@
...
+        <configuration>
+          <scalaVersion>${scala.version}</scalaVersion>
+          <scalaCompatVersion>${scala.binary.version}</scalaCompatVersion>
+          <args>
+            <arg>-Xmax-classfile-name</arg>
+            <arg>100</arg>
+            <arg>-Xlint</arg>
+          </args>
+        </configuration>
+      </plugin>
```
Comment (henriknyman) [inline community/cypher/frontend-3.4/pom.xml]: Is this intentionally added here?

File: community/cypher/pom.xml [hunk 24]
```diff
@@ -49,13 +54,43 @@
 
   <build>
     <plugins>
+
       <plugin>
         <groupId>net.alchim31.maven</groupId>
         <artifactId>scala-maven-plugin</artifactId>
+        <version>3.2.0</version>
```
Comment (henriknyman) [inline community/cypher/pom.xml]: Are these changes intentional?

File: community/cypher/aux-3.4/pom.xml [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+      <url>http://www.apache.org/licenses/LICENSE-2.0</url>
+    </license>
+  </licenses>
+
+  <build>
+    <plugins>
+
+      <plugin>
+        <groupId>net.alchim31.maven</groupId>
+        <artifactId>scala-maven-plugin</artifactId>
```
Comment (henriknyman) [inline community/cypher/aux-3.4/pom.xml]: Maybe verify that this is in sync with updates to scala-maven-plugin in other pom-files?

**Review by henriknyman — CHANGES_REQUESTED** [review]
I suspect that it suffers from some problematic rebase conflict resolution

File: community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/ExpressionTest.scala (hunk 11 above)
Comment (fickludd) [reply community/cypher/frontend-3.4/src/test/scala/org/neo4j/cypher/internal/frontend/v3_4/ast/ExpressionTest.scala:71]: This was actually correct, tests were moved to `SemanticAnalysisToolingTest`

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala (hunk 15 above)
Comment (fickludd) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/SemanticErrorAcceptanceTest.scala:362]: Yupp, I snook in a small improvement on the error message position, as that actually was easier to implement

**Review by henriknyman — COMMENTED** [review]
A lot of new folders have v3_3 in the path

**Review by henriknyman — APPROVED** [review]

**Pull request merged**